c/cert/src/rules/EXP32-C/DoNotAccessVolatileObjectWithNonVolatileReference.ql

```diff
@@ -0,0 +1,65 @@
+/**
+ * @id c/cert/do-not-access-volatile-object-with-non-volatile-reference
+ * @name EXP32-C: Do not access a volatile object through a nonvolatile reference
+ * @description If an an object defined with a volatile-qualified type is referred to with an lvalue
+ *              of a non-volatile-qualified type, the behavior is undefined.
+ * @kind problem
+ * @precision high
+ * @problem.severity error
+ * @tags external/cert/id/exp32-c
+ *       correctness
+ *       external/cert/obligation/rule
+ */
+
+import cpp
+import codingstandards.c.cert
+
+/**
+ * A `Cast` which converts from a pointer to a volatile-qualified type
+ * to a pointer to a non-volatile-qualified type.
+ */
+class CastFromVolatileToNonVolatileBaseType extends Cast {
+  CastFromVolatileToNonVolatileBaseType() {
+    this.getExpr().getType().(PointerType).getBaseType*().isVolatile() and
+    this.getActualType() instanceof PointerType and
+    not this.getActualType().(PointerType).getBaseType*().isVolatile()
+  }
+}
+
+/**
+ * An `AssignExpr` with an *lvalue* that is a pointer to a volatile base type and
+ * and *rvalue* that is not also a pointer to a volatile base type.
+ */
+class NonVolatileObjectAssignedToVolatilePointer extends AssignExpr {
+  NonVolatileObjectAssignedToVolatilePointer() {
+    this.getLValue().getType().(DerivedType).getBaseType*().isVolatile() and
+    not this.getRValue().getUnconverted().getType().(DerivedType).getBaseType*().isVolatile()
+  }
+
+  /**
+   * All `VariableAccess` expressions which are transitive successors of
+   * this `Expr` and which access the variable accessed in the *rvalue* of this `Expr`
+   */
+  Expr getASubsequentAccessOfAssignedObject() {
```

Comment:
I think it would be better to include this as part of the charpred for this class, e.g. `exists(VariableAccess va | ....`. I would then create an abstract super class to capture both, and provide a `string getMessage()` predicate so that we can simplify the `select`.

Reply:
Changes made.